### PR TITLE
[DROOLS-4618] Test Scenario: refactor ExpressionEvaluator implementations

### PR DESCRIPTION
**Pull request by danielezonca**
Changed `ExpressionEvaluator` input value to `String` (it should be possible to refactor `factMappingValue.getRawValue()` to `String` too but impact is big so will be done in the future)

@gitgabrio @yesamer @dupliaka 

https://issues.redhat.com/browse/DROOLS-4618

### CODE REVIEW

File: drools-scenario-simulation/drools-scenario-simulation-backend/src/main/java/org/drools/scenariosimulation/backend/expression/AbstractExpressionEvaluator.java [hunk 1]
```diff
@@ -67,6 +69,9 @@ protected boolean isStructuredInput(String className) {
     }
 
     protected Object convertResult(String rawString, String className, List<String> genericClasses) {
+        if(rawString == null) {
```
Comment (yesamer) [inline drools-scenario-simulation/drools-scenario-simulation-backend/src/main/java/org/drools/scenariosimulation/backend/expression/AbstractExpressionEvaluator.java]: Space between if and (

File: drools-scenario-simulation/drools-scenario-simulation-backend/src/main/java/org/drools/scenariosimulation/backend/expression/BaseExpressionOperator.java [hunk 2]
```diff
@@ -54,21 +56,20 @@ public String toString() {
...
+        private List<String> getValues(String rawValue) {
+            if (!match(rawValue).isPresent()) {
                 return Collections.emptyList();
             }
-            String rawValue = ((String) raw).trim();
             if (!rawValue.endsWith("]")) {
                 throw new IllegalArgumentException(new StringBuilder().append("Malformed expression: ").append(rawValue).toString());
             }
-            return Stream.of(rawValue.substring(1, ((String) raw).length() - 1).split(","))
+            return Stream.of(rawValue.substring(1, (rawValue).length() - 1).split(","))
```
Comment (yesamer) [inline drools-scenario-simulation/drools-scenario-simulation-backend/src/main/java/org/drools/scenariosimulation/backend/expression/BaseExpressionOperator.java]: Please remove parentheses on `rawValue` variable

**Review by yesamer — APPROVED** [review]
@danielezonca Excellent refactoring, I agree with you this should be considered as first step, the next one is to convert `factMappingValue.getRawValue()` from `Object` to `String` in a future ticket. We will have conflicts with my PR https://github.com/kiegroup/drools/pull/2691, but I think I can resolve them easily. So, I propose to review / test / merge this PR before mine.

File: drools-scenario-simulation/drools-scenario-simulation-backend/src/main/java/org/drools/scenariosimulation/backend/expression/MVELExpressionEvaluator.java [hunk 3]
```diff
@@ -44,29 +45,21 @@ public MVELExpressionEvaluator(ClassLoader classLoader) {
...
-    public Object evaluateLiteralExpression(String className, List<String> genericClasses, Object rawExpression) {
-        if (!(rawExpression instanceof String)) {
-            throw new IllegalArgumentException("Raw expression should be a String and not a '" + rawExpression.getClass().getCanonicalName() + "'");
-        }
-        Object expressionResult = compileAndExecute((String) rawExpression, Collections.emptyMap());
+    public Object evaluateLiteralExpression(String rawExpression, String className, List<String> genericClasses) {
+        Object expressionResult = compileAndExecute(rawExpression, Collections.emptyMap());
         Class<Object> requiredClass = loadClass(className, classLoader);
         if (expressionResult != null && !requiredClass.isAssignableFrom(expressionResult.getClass())) {
             throw new IllegalArgumentException("Cannot assign a '" + expressionResult.getClass().getCanonicalName() +
```
Comment (jomarko) [inline drools-scenario-simulation/drools-scenario-simulation-backend/src/main/java/org/drools/scenariosimulation/backend/expression/MVELExpressionEvaluator.java:65]: Minor thing, we miss ending apostrophe in the exception message.

File: drools-scenario-simulation/drools-scenario-simulation-backend/src/main/java/org/drools/scenariosimulation/backend/runner/AbstractRunnerHelper.java [hunk 4]
```diff
@@ -239,9 +239,9 @@ public void run(KieContainer kieContainer,
             ExpressionEvaluator expressionEvaluator = expressionEvaluatorFactory.getOrCreate(factMappingValue);
 
             try {
-                Object value = expressionEvaluator.evaluateLiteralExpression(factMapping.getClassName(),
-                                                                             factMapping.getGenericTypes(),
-                                                                             factMappingValue.getRawValue());
+                Object value = expressionEvaluator.evaluateLiteralExpression((String) factMappingValue.getRawValue(), factMapping.getClassName(),
```
Comment (jomarko) [inline drools-scenario-simulation/drools-scenario-simulation-backend/src/main/java/org/drools/scenariosimulation/backend/runner/AbstractRunnerHelper.java:242]: Can we please keep arguments of `evaluateLiteralExpression` one per line?

Comment (jomarko) [reply drools-scenario-simulation/drools-scenario-simulation-backend/src/main/java/org/drools/scenariosimulation/backend/runner/AbstractRunnerHelper.java:242]: Will we change also signature of `evaluateLiteralExpression` in future to eliminate this casting?

File: drools-scenario-simulation/drools-scenario-simulation-backend/src/main/java/org/drools/scenariosimulation/backend/runner/AbstractRunnerHelper.java [hunk 5]
```diff
@@ -301,7 +301,7 @@ protected ResultWrapper getResultWrapper(String className,
                                              Object resultRaw,
                                              Class<?> resultClass) {
         try {
-            boolean evaluationSucceed = expressionEvaluator.evaluateUnaryExpression(expectedResultRaw, resultRaw, resultClass);
+            boolean evaluationSucceed = expressionEvaluator.evaluateUnaryExpression((String) expectedResultRaw, resultRaw, resultClass);
```
Comment (jomarko) [inline drools-scenario-simulation/drools-scenario-simulation-backend/src/main/java/org/drools/scenariosimulation/backend/runner/AbstractRunnerHelper.java:304]: Will we change also signature of `evaluateUnaryExpression` in future to eliminate this casting?

**Review by jomarko — COMMENTED** [review]
See inline comments/questions.

File: drools-scenario-simulation/drools-scenario-simulation-backend/src/main/java/org/drools/scenariosimulation/backend/expression/MVELExpressionEvaluator.java (hunk 3 above)
Comment (danielezonca) [reply drools-scenario-simulation/drools-scenario-simulation-backend/src/main/java/org/drools/scenariosimulation/backend/expression/MVELExpressionEvaluator.java:65]: Good catch :)
Done

File: drools-scenario-simulation/drools-scenario-simulation-backend/src/main/java/org/drools/scenariosimulation/backend/runner/AbstractRunnerHelper.java (hunk 4 above)
Comment (danielezonca) [reply drools-scenario-simulation/drools-scenario-simulation-backend/src/main/java/org/drools/scenariosimulation/backend/runner/AbstractRunnerHelper.java:242]: We plan to remove the cast in the future changing `factMappingValue.rawValue` from `Object` to `String`

File: drools-scenario-simulation/drools-scenario-simulation-backend/src/main/java/org/drools/scenariosimulation/backend/runner/AbstractRunnerHelper.java (hunk 5 above)
Comment (danielezonca) [reply drools-scenario-simulation/drools-scenario-simulation-backend/src/main/java/org/drools/scenariosimulation/backend/runner/AbstractRunnerHelper.java:304]: Same as above, after the change of `factMappingValue.rawValue` from `Object` to `String` the cast will be no more needed

**Review by jomarko — APPROVED** [review]

**Pull request merged**